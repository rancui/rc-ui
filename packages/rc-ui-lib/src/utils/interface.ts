import type { CSSProperties } from 'react';

export interface BaseTypeProps {
  style?: CSSProperties;
  className?: string;
  children?: React.ReactNode;
}

/** 指定挂载的节点 */
export type TeleportType = HTMLElement | (() => HTMLElement) | null;
/** 带有displayName的组件 */
export type WithDisplayNameReactElement = React.ReactElement & { type: { displayName: string } };
