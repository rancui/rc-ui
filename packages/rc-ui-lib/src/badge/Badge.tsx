import React, { CSSProperties, useContext } from 'react';
import classnames from 'classnames';
import { BadgeProps } from './PropsType';
import { isDef, addUnit } from '../utils';
import { isNumeric } from '../utils/validate/number';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const Badge: React.FC<BadgeProps> = (props) => {
  const { content, max, dot, showZero, tag } = props;

  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('badge', prefixCls);

  const hasContent = () => {
    return isDef(content) && content !== '' && (showZero || +content !== 0);
  };

  const renderContent = () => {
    if (!dot && hasContent()) {
      if (isDef(max) && isNumeric(content?.toString()) && +content > +max) {
        return `${max}+`;
      }

      return content;
    }
    return null;
  };

  const getOffsetWithMinusString = (val: string) =>
    val.startsWith('-') ? val.replace('-', '') : `-${val}`;

  const renderBadge = () => {
    if (hasContent() || props.dot) {
      let style: CSSProperties = {
        background: props.color,
      };

      if (props.offset) {
        const [x, y] = props.offset;
        const { position } = props;
        const [offsetY, offsetX] = position.split('-') as ['top' | 'bottom', 'left' | 'right'];

        if (props.children) {
          if (typeof y === 'number') {
            style[offsetY] = addUnit(offsetY === 'top' ? y : -y);
          } else {
            style[offsetY] = offsetY === 'top' ? addUnit(y) : getOffsetWithMinusString(y);
          }

          if (typeof x === 'number') {
            style[offsetX] = addUnit(offsetX === 'left' ? x : -x);
          } else {
            style[offsetX] = offsetX === 'left' ? addUnit(x) : getOffsetWithMinusString(x);
          }
        } else {
          style.marginTop = addUnit(y);
          style.marginLeft = addUnit(x);
        }
      }

      if (!props.children) {
        style = { ...props.style, ...style };
      }
      return (
        <div
          className={classnames(
            {
              [props.className]: props.className && !props.children,
            },
            bem([props.position, { dot: props.dot, fixed: !!props.children }]),
          )}
          style={style}
        >
          {renderContent()}
        </div>
      );
    }
    return null;
  };

  if (props.children) {
    return React.createElement(
      tag,
      {
        className: classnames(bem('wrapper'), props.className),
        onClick: props.onClick,
        onTouchStart: props.onTouchStart,
        style: props.style,
      },
      props.children,
      renderBadge(),
    );
  }

  return renderBadge();
};

Badge.defaultProps = {
  tag: 'div',
  showZero: true,
  position: 'top-right',
};

export default Badge;
