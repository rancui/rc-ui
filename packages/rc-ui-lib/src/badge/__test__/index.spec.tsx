import React from 'react';
import { cleanup, render } from '@testing-library/react';
import Badge from '../index';
import Icon from '../../icon';
import { sleep } from '../../../tests/utils';

describe('Badge', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });
  it('should render nothing when content is empty string', () => {
    const { container } = render(<Badge content="" />);

    expect(container).toMatchSnapshot();
  });

  it('should render nothing when content is undefined', () => {
    const { container } = render(<Badge content="undefined" />);

    expect(container).toMatchSnapshot();
  });

  it('should render nothing when content is zero', () => {
    const { container } = render(<Badge content="0" />);

    expect(container).toMatchSnapshot();
  });

  it('should render className prop correctly', () => {
    const { container } = render(
      <Badge className="test-badge" content={<Icon name="success" className="badge-icon" />} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render content prop correctly', () => {
    const { container } = render(
      <Badge content={<Icon name="success" className="badge-icon" />}>
        <div className="child" />
      </Badge>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should change dot position when using offset prop', async () => {
    const { container, rerender } = render(
      <Badge dot offset={[2, 4]}>
        <div className="child" />
      </Badge>,
    );

    const style = getComputedStyle(container.querySelector('.rc-badge'));
    expect(style.top).toEqual('4px');
    expect(style.right).toEqual('-2px');

    rerender(
      <Badge dot offset={[-2, -4]}>
        <div className="child" />
      </Badge>,
    );

    await sleep(100);
    const style2 = getComputedStyle(container.querySelector('.rc-badge'));
    expect(style2.top).toEqual('-4px');
    expect(style2.right).toEqual('2px');
  });

  it('should change dot position when using offset prop with custom unit', async () => {
    const { container, rerender } = render(
      <Badge dot offset={['2rem', '4em']}>
        <div className="child" />
      </Badge>,
    );

    const style = getComputedStyle(container.querySelector('.rc-badge'));
    expect(style.top).toEqual('4em');
    expect(style.right).toEqual('-2rem');

    rerender(
      <Badge dot offset={['-2rem', '-4em']}>
        <div className="child" />
      </Badge>,
    );

    await sleep(10);
    const style2 = getComputedStyle(container.querySelector('.rc-badge'));

    expect(style2.top).toEqual('-4em');
    expect(style2.right).toEqual('2rem');
  });

  it('should change dot position when using offset prop without children', () => {
    const { container } = render(<Badge dot offset={[2, 4]} />);

    const badge = container.querySelector('.rc-badge');
    const style = getComputedStyle(badge);
    expect(style.marginTop).toEqual('4px');
    expect(style.marginLeft).toEqual('2px');
  });

  it('should not render zero when show-zero is false', async () => {
    const { container, rerender } = render(
      <Badge content="0">
        <div className="child" />
      </Badge>,
    );

    expect(container.querySelector('.rc-badge')).toBeTruthy();

    rerender(
      <Badge content="0" showZero={false}>
        <div className="child" />
      </Badge>,
    );

    await sleep(10);

    expect(container.querySelector('.rc-badge')).toBeFalsy();
  });

  it('can set custom max count', () => {
    const { container } = render(
      <Badge content={100} max={99}>
        news
      </Badge>,
    );
    expect(container.querySelector('.rc-badge').textContent).toBe('99+');
  });

  it('should change dot position when using offset prop and position is bottom-right', async () => {
    const { container, rerender } = render(
      <Badge dot offset={[2, '-4rem']} position="bottom-right">
        news
      </Badge>,
    );

    const badge = container.querySelector('.rc-badge');
    const style = getComputedStyle(badge);

    expect(style.bottom).toEqual('4rem');
    expect(style.right).toEqual('-2px');

    rerender(
      <Badge dot offset={['2rem', -4]} position="bottom-right">
        news
      </Badge>,
    );

    await sleep(10);

    const style2 = getComputedStyle(container.querySelector('.rc-badge'));

    expect(style2.bottom).toEqual('4px');
    expect(style2.right).toEqual('-2rem');
  });

  it('should change dot position when using offset prop and position is bottom-left', async () => {
    const { container, rerender } = render(
      <Badge dot offset={[2, '-4rem']} position="bottom-left">
        news
      </Badge>,
    );

    const badge = container.querySelector('.rc-badge');
    const style = getComputedStyle(badge);

    expect(style.bottom).toEqual('4rem');
    expect(style.left).toEqual('2px');

    rerender(
      <Badge dot offset={['2rem', 4]} position="bottom-left">
        news
      </Badge>,
    );

    await sleep(10);

    const style2 = getComputedStyle(container.querySelector('.rc-badge'));

    expect(style2.bottom).toEqual('-4px');
    expect(style2.left).toEqual('2rem');
  });

  it('should change dot position when using offset prop and position is top-left', async () => {
    const { container } = render(
      <Badge dot offset={[2, '-4rem']} position="top-left">
        news
      </Badge>,
    );

    const badge = container.querySelector('.rc-badge');
    const style = getComputedStyle(badge);

    expect(style.top).toEqual('-4rem');
    expect(style.left).toEqual('2px');
  });
});
