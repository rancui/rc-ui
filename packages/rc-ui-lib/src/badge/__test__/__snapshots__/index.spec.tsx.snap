// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render className prop correctly 1`] = `
<div>
  <div
    class="test-badge rc-badge rc-badge--top-right"
  >
    <i
      class="rc-badge__wrapper badge-icon van-icon van-icon-success"
    />
  </div>
</div>
`;

exports[`Badge should render content prop correctly 1`] = `
<div>
  <div
    class="rc-badge__wrapper"
  >
    <div
      class="child"
    />
    <div
      class="rc-badge rc-badge--top-right rc-badge--fixed"
    >
      <i
        class="rc-badge__wrapper badge-icon van-icon van-icon-success"
      />
    </div>
  </div>
</div>
`;

exports[`Badge should render nothing when content is empty string 1`] = `<div />`;

exports[`Badge should render nothing when content is undefined 1`] = `
<div>
  <div
    class="rc-badge rc-badge--top-right"
  >
    undefined
  </div>
</div>
`;

exports[`Badge should render nothing when content is zero 1`] = `
<div>
  <div
    class="rc-badge rc-badge--top-right"
  >
    0
  </div>
</div>
`;
